### PR TITLE
Expanding resample_to_latlon to be compatible with time- and/or depth-varying fields

### PR DESCRIPTION
**Pull request by dafyddstephenson**
`resample_to_latlon` previously only appeared to be compatible with two-dimensional data, so interpolating fields that varied in both depth and time seemingly required calling the function in a double loop. Have added two sets of lines: the first (before interpolating) detects any non-horizontal dimensions and stacks them into a single dimension, moving them to the final dimension of the data if necessary (where they are ignored by `pyresample.kd_tree.resample_nearest` and `pyresample.kd_tree.resample_custom`). The second moves the final, ignored dimension back to its original position and unstacks it into however many non-horizontal dimensions existed before interpolation. Rudimentary testing showed this was ~20 times faster for monthly averaged 3D 1992-2018 LLC90 fields than looping.

Two commits as the second was just to streamline the code slightly. Passed the `py.test` routines with only warnings.

### CODE REVIEW

File: .coveragerc
Comment (timothyas) [inline .coveragerc:15]: Is there a reason coveragerc got deleted?

File: .gitignore
Comment (timothyas) [inline .gitignore:3]: We don't want to ignore changes to .coveragerc

File: ecco_v4_py/resample_to_latlon.py
Comment (timothyas) [inline ecco_v4_py/resample_to_latlon.py:101]: Just for my own understanding, shouldn't `n_extra_dims>0` be enough? Is there a reason for the second condition?

File: .coveragerc
Comment (timothyas) [inline .coveragerc:15]: Why was .coveragerc deleted?

**Review by timothyas — CHANGES_REQUESTED** [review]
On a more technical level, I don't think .coveragerc should be removed. Could you let us know why you did that or otherwise bring it back and remove the line from gitignore?

More conceptually, this sounds awesome, but if we make a very large 2D field out of a 3D field, for example, won't we be interpolating between vertical levels in a weird way? Now where each 2D slice borders another, they will interpolate between each other based on an arbitrary 2D boundary. Unless I've missed something.

File: .coveragerc
Comment (dafyddstephenson) [reply .coveragerc:15]: To answer the more technical part, I think no: the pyresample routines expect a 1D array of "swath" data but allow for an extra dimension ("channels" in the pyresample documentation) along which no interpolation occurs, conventionally the final dimension. The trick here is just to take any dimensions along which no interpolation is desired (such as model level or time) and stack them into one final dimension to be treated as "channels". They get unstacked and put back in their original location after the interpolation. 

Regarding `.coveragerc`, I think it appeared after I ran the tests so I assumed this was some sort of test-specific output data without much thought. Thinking about it the `rc` would suggest otherwise. My mistake.

File: ecco_v4_py/resample_to_latlon.py
Comment (dafyddstephenson) [reply ecco_v4_py/resample_to_latlon.py:101]: Originally I had it with just `n_extra_dims>0` but things fail if there is a trivial extra dimension of length 1. The second condition is to check the extra dimension actually contains data/is "meaningful" in some way.

File: .gitignore
Comment (dafyddstephenson) [reply .gitignore:3]: Got it, will update.

File: .coveragerc
Comment (dafyddstephenson) [reply .coveragerc:15]: mistake, good catch.

**Comment by dafyddstephenson** [comment]
Thanks for the comments Tim. Lmk how is best to proceed - implement the changes, commit, and issue another pull request?

**Comment by timothyas** [comment]
Hey @dafyddstephenson thanks for all those comments. I'll need to take a closer look at the code, but you can keep updating your branch and it will be reflected in this PR - no need to close it or open anything new. You can get those files with something like

```bash
git checkout upstream/master .coveragerc 
```

And same for .gitignore assuming you called the ecco group branch upstream. Also for future ref I'd recommend making a new branch when you make a feature update - you may run into some funny business when you update your forked code after this PR gets merged. Happy to talk more about that if you like. 

Thanks again for the updates! I'll get back to you soon.

**Comment by timothyas** [comment]
Thanks @dafyddstephenson! I'm still just trying to make sure I understand this PR. Is all of this reshaping safe to do because it ultimately does the resampling by lat/lon coordinates (e.g. XC/YC values)? Then it would totally make sense to me. Sorry for my ignorance, I've never delved into the resampling, just trusted it blindly 😅🙈

**Comment by dafyddstephenson** [comment]
lol nothing to apologise for, that's exactly what it's doing. 

[Looking at `pyresample.kd_tree.resample_nearest`](https://pyresample.readthedocs.io/en/latest/howtos/swath.html#resample-nearest) it takes a "Swath Definition" object (1D track of lons and lats) along with data (np array) along that swath. 

ECCOv4-py is currently providing a 1D swath object along with 3D data, [but pyresample just flattens that to 1D anyway in this `elif`](https://github.com/pytroll/pyresample/blob/main/pyresample/kd_tree.py):
```
if data.ndim > 2 and data.shape[0] * data.shape[1] == valid_input_index.size:
        data = data.reshape(data.shape[0] * data.shape[1], data.shape[2])
        elif data.shape[0] != valid_input_index.size:
        data = data.ravel()
```
What the change in this PR does is reshape the data into 2D so that it already looks like something pyresample expects: 1D of "swath" (lat and lon) and a second dimension of "channels", so it skips this `if` block entirely. The "channels" we're providing are just any dimensions that aren't lon and lat, repacked into a single dimension, and unpacked again at the end.

Without the change, if space or time is left in the data given to `resample_to_latlon`, these also get flattened in the above `ravel()` call, and the data and swath are mismatched, leading to `ValueError: Mismatch between geometry and dataset` in `kd_tree`, e.g. when trying the below:

```
GDS=xr.open_dataset('ECCO-GRID.nc')
lo,la,_,_,R=ecco.resample_to_latlon(GDS.XC,GDS.YC,GDS.hFacC,-90,90,1,-180,180,1,mapping_method='nearest_neighbor');
```

Thanks again for revisiting this :)

**Comment by timothyas** [comment]
Ok great, that makes sense. Thank you for the explanation @dafyddstephenson. I think this is ready to be merged, but I think @ifenty or @owang01 should take a look before merging. Could one of you guys check this out? I think it will be useful.

File: ecco_v4_py/resample_to_latlon.py
Comment (dafyddstephenson) [reply ecco_v4_py/resample_to_latlon.py:110]: There are 2 additional dimensions here (`n_extra_dims`) so
`extra_dim_shape` would be `orig_field.shape[:2]` or `[5,3,13,90,90][:2]` , so `[5,3]`

`np.prod(extra_dim_shape)` then would be `np.prod([5,3])` and `np.prod(horiz_dim_shape)` would be `np.prod([13,90,90])` resulting in

`new_shape` being `np.hstack([5*3,13*90*90])`, or np.array([15,105300])

Hopefully that addresses the question? sorry if not - happy to try again

Comment (timothyas) [reply ecco_v4_py/resample_to_latlon.py:110]: I should've resolved this a couple days ago when I looked at it again. For some reason last year when you opened this I didn't understand it, but now it's pretty clear. So yeah all good, thanks @dafyddstephenson

Comment (dafyddstephenson) [reply ecco_v4_py/resample_to_latlon.py:110]: oh, just saw the timestamp on it - my bad! 😅

**Pull request merged**
